src/binary/buffer.rs

```diff
@@ -0,0 +1,90 @@
+use crate::common::{Bytes32, Position};
+use core::mem::size_of;
+
+const BUFFER_SIZE: usize = size_of::<Position>() + size_of::<Bytes32>();
+pub type Buffer = [u8; BUFFER_SIZE];
+pub const DEFAULT_BUFFER: &Buffer = &[0; BUFFER_SIZE];
+
+struct Schema {}
+
+impl Schema {
+    const fn position_offset() -> isize {
+        0
+    }
+
+    const fn hash_offset() -> isize {
+        Self::position_offset() + size_of::<Position>() as isize
+    }
+}
+
+pub struct ReadView<'a> {
+    buffer: &'a Buffer,
+}
+
+impl<'a> ReadView<'a> {
+    fn buffer(&self) -> &Buffer {
+        self.buffer
+    }
+
+    unsafe fn position_ptr(&self) -> *const Position {
+        let offset = Schema::position_offset();
+        let position = self.buffer().as_ptr().offset(offset) as *const Position;
+        position
+    }
+
+    unsafe fn hash_ptr(&self) -> *const Bytes32 {
+        let offset = Schema::hash_offset();
+        let hash = self.buffer().as_ptr().offset(offset) as *const Bytes32;
+        hash
+    }
+
+    pub fn new(buffer: &'a Buffer) -> Self {
+        Self { buffer }
+    }
+
+    pub fn position(&self) -> Position {
+        // SAFETY: position_ptr is guaranteed to point to a valid Position
+        unsafe { *self.position_ptr() }
```

Comment:
Unfamiliar but is this unsafe code absolutely necessary?

Reply:
Here, I am dereferencing a pointer to a `Position`, where the pointer points to some bytes in the buffer reserved for a `Position` instance. Dereferencing a pointer is considered unsafe by definition, so in this context, it is necessary. 

The pattern I am trying to achieve here is to simply allocate the complete buffer of memory first, and then use contiguous slices of the memory for specific objects (i.e. `Position`). This way, whenever I am accessing or mutating the specific object, I always know that the memory I am operating on is inside this buffer. I could investigate other ways of achieving this pattern without unsafe code, but that will likely require finding ways to serialize/deserialize the objects into/from bytes and performing extra allocations and buffer writes.

Reply:
The closest alternative I can think of that would avoid writing any unsafe code ourselves is to use a zero-copy framework like rkiv. Although since these types are not very complex (no heap-based types etc) it could be overkill and add a bit of dependency bloat.

This function in particular seems like it could be unsound to me, because it returns `Position` instead of `&Position`. This means that if the underlying object is moved (which is allow when there are no outstanding leases), we could end up with a dangling pointer.

Reply:
In this case, the function actually performs a sort of copy-on-read, since we are always returning a new instance of `Position` (`Position` is simply a wrapper around `u64`, so we can implement copy). The public interfaces for the `ReadView` and `WriteView` structs do not expose any pointers - only references (i.e. `&Bytes32`) or newly allocated instances (i.e. `Position`). 

In practice, the buffer always outlives any buffer views. We create a new view on the buffer every time we need to access the buffer via the parent `Node` instance. Specifically, a new view is created during the scope of a `Node` getter or setter call, and the internal pointer arithmetic is recalculated at that time. Furthermore, if any references are returned from a view, we cannot move the buffer as long as that reference is held.

Reply:
ah gotcha, might be worth adding a note to the safety section about how position is copied here.